Use a callback with Motion History Reporter

p2 opened the pull request:
- New, additional method to collect motion data and be notified on a callback
- Preserves the current method with a notification center notification but uses the callback internally
- Add Swift nullable annotations

p2 committed:
Use a callback with Motion History Reporter

- New, additional method to collect motion data and be notified on a callback
- Preserves the current method with a notification center notification but uses the callback internally
- Add Swift nullable annotations

YuanZhu-apple commented on APCAppCore/APCAppCore/Library/APCMotionHistoryReporter.h:
```diff
@@ -34,18 +35,23 @@
++ (APCMotionHistoryReporter *)sharedInstance;
 
+/** Start motion history collection. You will need to listen to the "" notification and call the shared instance's `retrieveMotionReport`. */
+- (void)startMotionCoProcessorDataFrom:(NSDate *)startDate andEndDate:(NSDate *)endDate andNumberOfDays:(NSInteger)numberOfDays;
```
"APCMotionHistoryReporterDoneNotification" is missed in the comments.

YuanZhu-apple commented on APCAppCore/APCAppCore/Library/APCMotionHistoryReporter.m:
```diff
@@ -72,37 +74,50 @@ +(APCMotionHistoryReporter *) sharedInstance {
+
+- (void)startMotionCoProcessorDataFrom:(NSDate * __nonnull)startDate andEndDate:(NSDate * __nonnull)endDate andNumberOfDays:(NSInteger)numberOfDays callback:(APCMotionHistoryReporterCallback __nonnull)callback{
+	if (_doneCallback) {
+		callback(nil, [NSError errorWithDomain:@"APCAppCoreErrorDomain" code:51 userInfo:@{NSLocalizedDescriptionKey: @"Motion History Reporter is already processing motion history, wait for it to complete"}]);
```
What does `51` stand for? Maybe  add a little bit documentation?

YuanZhu-apple commented on APCAppCore/APCAppCore/Library/APCMotionHistoryReporter.m:
```diff
@@ -72,37 +74,50 @@ +(APCMotionHistoryReporter *) sharedInstance {
+}
+
+- (void)startMotionCoProcessorDataFrom:(NSDate * __nonnull)startDate andEndDate:(NSDate * __nonnull)endDate andNumberOfDays:(NSInteger)numberOfDays callback:(APCMotionHistoryReporterCallback __nonnull)callback{
+	if (_doneCallback) {
```
What if someone passed in a nil callback? 
Then it could be ran over by second call, and also won't get a callback/notification. 
Could we throw an exception on receiving a nil callback?

p2 committed:
Early exit if num days = 0. Address PR comments.

p2 commented:
I've addressed the comments and have another commit ready that adjusts all the whitespace. Holding back with that to not make reviewing harder for you, let me know when I should push.

I would have liked to add better method documentation, but I've been unable to figure out what the `numberOfDays` param is supposed to do. The only number that makes sense to me is **1**.

YuanZhu-apple commented:
@chrisinsfo This looks good to me. Can you review this as well?

ghost commented:
@YuanZhu-apple We have no problems with this.

YuanZhu-apple commented:
OK, mereged
